### PR TITLE
implement more drop stuff

### LINKED ISSUE #79: Box is deallocated before drop impl is run
```rust
struct Foo(Box<u8>);
impl Drop for Foo {
    fn drop(&mut self) {
        *self.0 = 255; // error: dangling pointer dereferenced
    }
}

fn main() {
    drop(Foo(Box::new(0)));
}
```

Boxes are immediately deallocated by the drop terminator, where as drop implementations are pushed to the stack, https://github.com/solson/miri/blob/master/src/interpreter/terminator/mod.rs#L531
I'm not sure how to solve this, other than maybe adding a separate type of frame just to deallocate the box.

### PR DESCRIPTION
**Pull request by oli-obk**
I haven't touched `TyClosure` and `TyAnon`, not even sure where to start for those...

`TyAnon` could probably be solved by a `MirPass` that replaces all of them with the actual type

Fixes #79

### CODE REVIEW

**Comment by eddyb** [comment]
You can't see `TyAnon` in a monomorphized type.

**Comment by solson** [comment]
I think a `TyClosure` just drops its fields like any tuple, array, or univariant.

File: src/interpreter/terminator/mod.rs [hunk 1]
```diff
@@ -577,25 +614,57 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
+
+    fn drop_fields<
+        I: Iterator<Item=(Ty<'tcx>, ty::layout::Size)>,
+    >(
+        &mut self,
+        mut fields: I,
+        lval: Lvalue<'tcx>,
+        drop: &mut Vec<(DefId, Value, &'tcx Substs<'tcx>)>,
+    ) -> EvalResult<'tcx, ()> {
+        // FIXME: some aggregates may be represented by PrimVal::Pair
```
Comment (solson) [inline src/interpreter/terminator/mod.rs]: `s/PrimVal::Pair/Value::ByValPair/`

File: src/interpreter/terminator/mod.rs [hunk 2]
```diff
@@ -577,25 +614,57 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
+        let (adt_ptr, extra) = match self.force_allocation(lval)? {
```
Comment (solson) [inline src/interpreter/terminator/mod.rs]: Use `.to_ptr_and_extra()` instead of this match.

File: src/interpreter/terminator/intrinsics.rs [hunk 3]
```diff
@@ -123,9 +123,20 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
                 let ty = substs.type_at(0);
-                let ptr = arg_vals[0].read_ptr(&self.memory)?;
+                let ptr_ty = self.tcx.mk_mut_ptr(ty);
+                let lvalue = match self.follow_by_ref_value(arg_vals[0], ptr_ty)? {
+                    Value::ByRef(_) => bug!("follow_by_ref_value returned ByRef"),
+                    Value::ByVal(ptr) => Lvalue::from_ptr(ptr.expect_ptr("drop_in_place first arg not a pointer")),
+                    Value::ByValPair(ptr, extra) => Lvalue::Ptr {
+                        ptr: ptr.expect_ptr("drop_in_place first arg not a pointer"),
+                        extra: match extra.try_as_ptr() {
+                            Some(vtable) => LvalueExtra::Vtable(vtable),
```
Comment (solson) [inline src/interpreter/terminator/intrinsics.rs]: What if we're dropping something like `Box<[T]>`? Do we reach this `Vtable` line then?

File: src/interpreter/terminator/mod.rs [hunk 4]
```diff
@@ -525,18 +526,53 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
+                                ptr: ptr,
+                                extra: extra,
+                            },
+                            contents_ty,
+                            drop,
+                        )?;
+                    },
+                }
+                let box_free_fn = self.tcx.lang_items.box_free_fn().expect("no box_free lang item");
+                let substs = self.tcx.intern_substs(&[Kind::from(contents_ty)]);
```
Comment (oli-obk) [inline src/interpreter/terminator/mod.rs]: This fixes #79

**Pull request merged**